Illegal character in path when classpath contains entry with URL-encoded space

Yes, it looks like another Windows URI issue.

Running tests from IntelliJ I get the error below. It is only logged as a warning, and the tests still run successfully. However, with some ~20 instances of that stack trace, it pollutes the logs quite a bit.

Interesting, too, that the path seems to be URL encoded initially:

`Cannot parse URL classpath entry [...]/Program%20Files/[...]`
`[...] Illegal character in path at index 11: [...]/Program Files/[...]`

ArchUnit version `0.14.1`, Junit 5.

```
13:58:48.982 [main] WARN  com.tngtech.archunit.core.importer.UrlSource$From - Cannot parse URL classpath entry file:/C:/Program%20Files/JetBrains/IntelliJ%20IDEA%202019.2.3/lib/idea_rt.jar
java.lang.IllegalArgumentException: Illegal character in path at index 11: /C:/Program Files/JetBrains/IntelliJ IDEA 2019.2.3/lib/idea_rt.jar
	at java.net.URI.create(URI.java:852) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI.resolve(URI.java:1036) ~[?:1.8.0_201-1-ojdkbuild]
	at com.tngtech.archunit.core.importer.UrlSource$From.parseUrl(UrlSource.java:156) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.UrlSource$From.parseManifestClasspathEntry(UrlSource.java:144) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.UrlSource$From.readClasspathEntriesFromManifest(UrlSource.java:115) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.UrlSource$From.readClasspathUriEntriesFromManifests(UrlSource.java:99) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.UrlSource$From.readClasspathEntriesFromManifests(UrlSource.java:91) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.UrlSource$From.classPathSystemProperties(UrlSource.java:85) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.LocationResolver$Legacy.resolveClassPath(LocationResolver.java:38) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.Locations.getLocationsOf(Locations.java:112) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.core.importer.Locations.ofPackage(Locations.java:74) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache$RequestedLocations$Specific.locationsOf(ClassCache.java:220) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache$RequestedLocations$Specific.getLocationsOfPackages(ClassCache.java:187) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache$RequestedLocations$Specific.<init>(ClassCache.java:177) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache$RequestedLocations$Specific.<init>(ClassCache.java:170) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache$RequestedLocations.by(ClassCache.java:234) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ClassCache.getClassesToAnalyzeFor(ClassCache.java:76) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ArchUnitTestDescriptor.lambda$createChildren$1(ArchUnitTestDescriptor.java:82) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.thirdparty.com.google.common.base.Suppliers$MemoizingSupplier.get(Suppliers.java:120) ~[archunit-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ArchUnitTestDescriptor$ArchUnitRuleDescriptor.execute(ArchUnitTestDescriptor.java:159) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at com.tngtech.archunit.junit.ArchUnitTestDescriptor$ArchUnitRuleDescriptor.execute(ArchUnitTestDescriptor.java:142) ~[archunit-junit5-engine-0.14.1.jar:0.14.1]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:112) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_201-1-ojdkbuild]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.Iterator.forEachRemaining(Iterator.java:116) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_201-1-ojdkbuild]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418) ~[?:1.8.0_201-1-ojdkbuild]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120) ~[junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66) [junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108) [junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79) [junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55) [junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43) [junit-platform-engine-1.1.0.jar:1.1.0]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170) [junit-platform-launcher-1.1.0.jar:1.1.0]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154) [junit-platform-launcher-1.1.0.jar:1.1.0]
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90) [junit-platform-launcher-1.1.0.jar:1.1.0]
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:69) [junit5-rt.jar:?]
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47) [junit-rt.jar:?]
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242) [junit-rt.jar:?]
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70) [junit-rt.jar:?]
Caused by: java.net.URISyntaxException: Illegal character in path at index 11: /C:/Program Files/JetBrains/IntelliJ IDEA 2019.2.3/lib/idea_rt.jar
	at java.net.URI$Parser.fail(URI.java:2848) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI$Parser.checkChars(URI.java:3021) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI$Parser.parseHierarchical(URI.java:3105) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI$Parser.parse(URI.java:3063) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI.<init>(URI.java:588) ~[?:1.8.0_201-1-ojdkbuild]
	at java.net.URI.create(URI.java:850) ~[?:1.8.0_201-1-ojdkbuild]
	... 63 more
```

--- a/archunit/src/main/java/com/tngtech/archunit/core/importer/UrlSource.java
+++ b/archunit/src/main/java/com/tngtech/archunit/core/importer/UrlSource.java
@@ -153,7 +153,7 @@ interface UrlSource extends Iterable<URL> {
 
         private static Optional<URI> parseUrl(Path parent, String classpathUrlEntry) {
             try {
-                return Optional.of(convertToJarUrlIfNecessary(parent.toUri().resolve(URI.create(classpathUrlEntry).getSchemeSpecificPart())));
+                return Optional.of(convertToJarUrlIfNecessary(parent.toUri().resolve(URI.create(classpathUrlEntry).getRawSchemeSpecificPart())));
             } catch (Exception e) {
                 LOG.warn("Cannot parse URL classpath entry " + classpathUrlEntry, e);
                 return Optional.absent();

--- a/archunit/src/test/java/com/tngtech/archunit/core/importer/UrlSourceTest.java
+++ b/archunit/src/test/java/com/tngtech/archunit/core/importer/UrlSourceTest.java
@@ -109,6 +109,20 @@ public class UrlSourceTest {
     }
 
     @Test
+    public void handles_jar_uri_with_spaces() throws Exception {
+        File folderWithSpaces = temporaryFolder.newFolder("folder with spaces");
+        File folder = temporaryFolder.newFolder();
+
+        WrittenJarFile jarInFolderWithSpaces = writeJarWithManifestClasspathAttribute(folderWithSpaces, "folder-with-spaces");
+        WrittenJarFile parentJar = writeJarWithManifestClasspathAttribute(folder, "parent", ManifestClasspathEntry.absoluteUrl(jarInFolderWithSpaces.path));
+
+        System.setProperty(JAVA_CLASS_PATH_PROP, parentJar.path.toString());
+        UrlSource urls = UrlSource.From.classPathSystemProperties();
+
+        assertThat(urls).containsAll(concat(parentJar.getExpectedClasspathUrls(), jarInFolderWithSpaces.getExpectedClasspathUrls()));
+    }
+
+    @Test
     public void recursively_resolves_classpath_attributes_in_manifests() throws Exception {
         File folder = temporaryFolder.newFolder();
         WrittenJarFile grandChildOne = writeJarWithManifestClasspathAttribute(folder, subPath("grandchild", "one"));

